File: bot.js
const TelegramBot = require('node-telegram-bot-api');

const token = '7480102557:AAHqayZi6LXwBVrwK6babLtOgie454WYZJg'; // Замените на свой токен
const bot = new TelegramBot(token, { polling: true });

const timeMapping = {
    '5_minutes': 5 * 60 * 1000,
    '15_minutes': 15 * 60 * 1000,
    '30_minutes': 30 * 60 * 1000,
    '1_hour': 60 * 60 * 1000,
    '2_hours': 2 * 60 * 60 * 1000,
    '3_hours': 3 * 60 * 60 * 1000,
};

const scheduledMessages = {};

bot.onText(/\/start/, (msg) => {
    const chatId = msg.chat.id;
    bot.sendMessage(chatId, 'Добро пожаловать! Чтобы установить напоминание, выберите день.', {
        reply_markup: {
            inline_keyboard: [
                [{ text: 'Сегодня', callback_data: 'today' }],
                [{ text: 'Завтра', callback_data: 'tomorrow' }],
                [{ text: 'Выбрать дату', callback_data: 'choose_date' }],
            ],
        },
    });
});

bot.on('callback_query', async (query) => {
    await bot.answerCallbackQuery(query.id);
    const chatId = query.message.chat.id;

    if (query.data === 'today') {
        bot.sendMessage(chatId, 'Через сколько поставить напоминание?', {
            reply_markup: {
                inline_keyboard: [
                    [{ text: '5 минут', callback_data: '5_minutes' }],
                    [{ text: '15 минут', callback_data: '15_minutes' }],
                    [{ text: '30 минут', callback_data: '30_minutes' }],
                    [{ text: '1 час', callback_data: '1_hour' }],
                    [{ text: '2 часа', callback_data: '2_hours' }],
                    [{ text: '3 часа', callback_data: '3_hours' }],
                    [{ text: 'Утром', callback_data: 'morning' }],
                    [{ text: 'Днём', callback_data: 'day' }],
                    [{ text: 'Вечером', callback_data: 'evening' }],
                    [{ text: 'Перед сном', callback_data: 'before_sleep' }],
                ],
            },
        });
    } else if (query.data === 'tomorrow' || query.data === 'choose_date') {
        const message = query.data === 'choose_date' ? 'Напишите дату в формате дд:мм' : 'Выберите время для напоминания:';
        bot.sendMessage(chatId, message, {
            reply_markup: {
                inline_keyboard: [
                    [{ text: 'Утром', callback_data: 'morning' }],
                    [{ text: 'Днём', callback_data: 'day' }],
                    [{ text: 'Вечером', callback_data: 'evening' }],
                    [{ text: 'Перед сном', callback_data: 'before_sleep' }],
                    [{ text: 'Задать время', callback_data: 'set_time' }],
                ],
            },
        });
    } else if (query.data === 'set_time') {
        bot.sendMessage(chatId, 'Напишите время в формате чч:мм');
        bot.on('message', (msg) => handleTimeSetting(msg, chatId));
    } else {
        handleTimeSelection(chatId, query.data);
    }
});

function handleTimeSetting(msg, chatId) {
    const time = msg.text.split(':');
    if (time.length === 2) {
        const hours = parseInt(time[0]);
        const minutes = parseInt(time[1]);
        if (!isNaN(hours) && !isNaN(minutes) && hours >= 0 && hours < 24 && minutes >= 0 && minutes < 60) {
            const waitTime = new Date().setHours(hours, minutes, 0, 0) - Date.now();
            if (waitTime > 0) {
                scheduleMessage(chatId, msg.chat.last_text, waitTime);
                bot.sendMessage(chatId, `Напоминание установлено на ${hours}:${minutes}.`);
            } else {
                bot.sendMessage(chatId, 'Выберите время в будущем.');
            }
        } else {
            bot.sendMessage(chatId, 'Неверный формат времени. Попробуйте ещё раз в формате чч:мм');
        }
    } else {
        bot.sendMessage(chatId, 'Неверный формат времени. Попробуйте ещё раз в формате чч:мм');
    }
}

function handleTimeSelection(chatId, timeFrame) {
    let waitTime;
    switch (timeFrame) {
        case 'morning':
            waitTime = 8 * 60 * 60 * 1000; // 8:00
            break;
        case 'day':
            waitTime = 12 * 60 * 60 * 1000; // 12:00
            break;
        case 'evening':
            waitTime = 18 * 60 * 60 * 1000; // 18:00
            break;
        case 'before_sleep':
            waitTime = 22 * 60 * 60 * 1000; // 22:00
            break;
        default:
            waitTime = timeMapping[timeFrame];
    }

    if (waitTime) {
        bot.sendMessage(chatId, `Напоминание установлено через ${waitTime / 60000} минут.`);
        scheduleMessage(chatId, 'Ваше напоминание!', waitTime);
    }
}

function scheduleMessage(chatId, message, waitTime) {
    setTimeout(() => {
        bot.sendMessage(chatId, message);
        // Если вам нужно удалить сообщение из scheduledMessages, добавьте здесь логику
    }, waitTime);
}

File: bot.py
import logging
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import ApplicationBuilder, CommandHandler, ContextTypes, MessageHandler, filters, CallbackQueryHandler
import threading
import time
from datetime import datetime, timedelta

# Включаем логирование
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                    level=logging.INFO)

logger = logging.getLogger(__name__)

# Токен вашего бота
TOKEN = '7480102557:AAHqayZi6LXwBVrwK6babLtOgie454WYZJg'

# Этот словарь хранит отложенные сообщения
scheduled_messages = {}


class Reminder:
    def __init__(self, message, chat_id, date_time, description=None):
        self.message = message
        self.chat_id = chat_id
        self.date_time = date_time
        self.description = description


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    await update.message.reply_text('Привет! Используйте /set_reminder, чтобы настроить напоминание.')


async def set_reminder(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    await update.message.reply_text('Какое сообщение вы хотите получить?')
    return 'GET_MESSAGE'


async def get_message(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    context.user_data['message'] = update.message.text
    await update.message.reply_text('Есть ли у этого сообщения пояснение? (напишите "нет", если пропустите)')
    return 'GET_DESCRIPTION'


async def get_description(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    description = update.message.text if update.message.text.lower() != 'нет' else None
    context.user_data['description'] = description

    keyboard = [
        [InlineKeyboardButton("1 час", callback_data='1_hour')],
        [InlineKeyboardButton("2 часа", callback_data='2_hours')],
        [InlineKeyboardButton("3 часа", callback_data='3_hours')]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    await update.message.reply_text('Когда вы хотите отправить напоминание?', reply_markup=reply_markup)
    return 'GET_TIME'


async def get_time(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    await query.answer()
    
    time_mapping = {
        '1_hour': timedelta(hours=1),
        '2_hours': timedelta(hours=2),
        '3_hours': timedelta(hours=3)
    }

    wait_time = time_mapping[query.data]
    message = context.user_data['message']
    description = context.user_data.get('description', None)
    reminder_time = datetime.now() + wait_time
    
    scheduled_messages[update.effective_chat.id] = Reminder(message, update.effective_chat.id, reminder_time, description)

    await query.edit_message_text(text=f'Напоминание установлено! Сообщение: "{message}", отправка через {wait_time}.')
    return -1  # Завершаем разговор


async def send_scheduled_messages():
    while True:
        now = datetime.now()
        for chat_id, reminder in list(scheduled_messages.items()):
            if reminder.date_time <= now:
                await updater.bot.send_message(chat_id=chat_id, text=f'Напоминание: {reminder.message}')
                if reminder.description:
                    await updater.bot.send_message(chat_id=chat_id, text=f'Описание: {reminder.description}')
                del scheduled_messages[chat_id]  # Удаляем сообщение после отправки
        await asyncio.sleep(60)  # Проверяем каждую минуту


def main():
    # Создаём объект приложения
    global updater
    updater = ApplicationBuilder().token(TOKEN).build()

    # Обработчики команд
    updater.add_handler(CommandHandler("start", start))
    updater.add_handler(CommandHandler("set_reminder", set_reminder))

    # Обработчики состояний
    updater.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, get_message))
    updater.add_handler(CallbackQueryHandler(get_time))

    # Запускаем поток для отправки запланированных сообщений
    threading.Thread(target=send_scheduled_messages, daemon=True).start()

    # Запуск бота
    updater.run_polling()


if __name__ == '__main__':
    main()
